src/realm/object-store/sectioned_results.hpp

```diff
@@ -0,0 +1,208 @@
+
+    /// Return a new instance SectionedResults that uses a snapshot of the underlying `Results`.
+    /// The section key callback parameter will never be invoked.
+    SectionedResults snapshot();
+
+    bool is_valid() const;
+    bool is_frozen() const;
+
+private:
+    friend class Results;
+    /// SectionedResults should not be created directly and should only be instantiated from `Results`.
+    SectionedResults(Results results, SectionKeyFunc section_key_func);
+    SectionedResults(Results results, Results::SectionedResultsOperator op, util::Optional<StringData> prop_name);
+    uint_fast64_t get_content_version();
+
+    friend struct SectionedResultsNotificationHandler;
+    void calculate_sections_if_required();
+    void calculate_sections();
+    bool has_performed_initial_evalutation = false;
+    NotificationToken add_notification_callback_for_section(Mixed section_key,
+                                                            SectionedResultsNotificatonCallback callback,
+                                                            KeyPathArray key_path_array = {});
+
+    friend class realm::ResultsSection;
+    Results m_results;
+    SectionKeyFunc m_callback;
+    std::unordered_map<Mixed, Section> m_sections;
+    // Stores the Key, Section Index of the previous section
+    // so we can efficiently calculate the collection change set.
+    std::unordered_map<Mixed, size_t> m_previous_key_to_index_lookup;
+    // Returns the key of the previous section from its index.
+    std::unordered_map<size_t, Mixed> m_prev_section_index_to_key;
+    // Returns the key of the current section from its index.
+    std::vector<Mixed> m_ordered_section_keys;
+    // By passing the index of the object from the underlying `Results`,
+    // this will give a pair with the section index of the object, and the position of the object in that section.
+    // This is used for parsing the indices in CollectionChangeSet to section indices.
+    std::vector<std::pair<size_t, size_t>> m_row_to_index_path;
+    // BinaryData & StringData types require a buffer to hold deep
+    // copies of the key values for the lifetime of the sectioned results.
+    // This is due to the fact that such values can reference the memory address of the value in the realm.
+    // We can not rely on that because it would not produce stable keys.
+    // So we perform a deep copy to produce stable key values that will not change if the realm is modified.
+    // The buffer will purge keys that are no longer used in the case that the `calculate_sections` method runs.
+    // We need to hold the string value in a shared_ptr so that if multiple notification handlers are attached
+    // to this `SectionedResults` the copy of the key will not be discarded during `calculate_sections` as each
+    // notification handler will hold a reference to the shared_ptr.
+    std::unordered_map<Mixed, std::shared_ptr<std::string>> m_previous_str_buffers, m_current_str_buffers;
```

Comment:
Using a map and a shared_ptr is almost certainly slower and more memory intensive than just storing a copy of the key in the notification handler.